File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/crypto/keystore/KeystoreKeyProvider.java
package org.hyperledger.iroha.android.crypto.keystore;

import java.security.KeyPair;
import java.security.MessageDigest;
import java.util.Arrays;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import org.hyperledger.iroha.android.IrohaKeyManager;
import org.hyperledger.iroha.android.KeyManagementException;
import org.hyperledger.iroha.android.crypto.KeyProviderMetadata;
import org.hyperledger.iroha.android.crypto.keystore.attestation.AttestationResult;
import org.hyperledger.iroha.android.crypto.keystore.attestation.AttestationVerificationException;
import org.hyperledger.iroha.android.crypto.keystore.attestation.AttestationVerifier;

/**
 * {@link IrohaKeyManager.KeyProvider} backed by an Android Keystore backend.
 *
 * <p>The backend is supplied through {@link KeystoreBackend}; desktop JVM builds can rely on fake
 * implementations for tests while Android builds will provide a real backend that bridges to
 * {@code android.security.keystore.KeyStore} / {@code KeyGenParameterSpec}. Support for StrongBox
 * and discrete secure elements will land by swapping the backend at runtime depending on device
 * capabilities.
 */
public final class KeystoreKeyProvider implements IrohaKeyManager.KeyProvider {

  private static final byte[] NO_CHALLENGE = new byte[0];

  private final KeystoreBackend backend;
  private final KeyGenParameters parameters;
  private final IrohaKeyManager.KeySecurityPreference defaultPreference;
  private final Map<CacheKey, KeyAttestation> attestationCache = new ConcurrentHashMap<>();

  public KeystoreKeyProvider(final KeystoreBackend backend, final KeyGenParameters parameters) {
    this(backend, parameters, null);
  }

  private KeystoreKeyProvider(
      final KeystoreBackend backend,
      final KeyGenParameters parameters,
      final IrohaKeyManager.KeySecurityPreference defaultPreference) {
    this.backend = Objects.requireNonNull(backend, "backend");
    this.parameters = Objects.requireNonNull(parameters, "parameters");
    this.defaultPreference = defaultPreference;
  }

  @Override
  public Optional<KeyPair> load(final String alias) throws KeyManagementException {
    return backend.load(alias);
  }

  @Override
  public KeyPair generate(final String alias) throws KeyManagementException {
    if (defaultPreference != null) {
      return generateWithOutcome(alias, defaultPreference).keyPair();
    }
    evictAttestations(alias);
    if (parameters.requireStrongBox() && !backend.metadata().strongBoxBacked()) {
      throw new KeyManagementException("StrongBox required but backend is not StrongBox-capable");
    }
    final KeyGenerationResult result = backend.generate(alias, parameters);
    final org.hyperledger.iroha.android.crypto.KeyGenerationOutcome outcome =
        outcomeFor(alias, result.keyPair());
    if (parameters.requireStrongBox()
        && outcome.route()
            != org.hyperledger.iroha.android.crypto.KeyGenerationOutcome.Route.STRONGBOX) {
      throw new KeyManagementException(
          "StrongBox required but backend produced a weaker security level");
    }
    return outcome.keyPair();
  }

  @Override
  public KeyPair generate(
      final String alias, final org.hyperledger.iroha.android.IrohaKeyManager.KeySecurityPreference preference)
      throws KeyManagementException {
    return generateWithOutcome(alias, preference).keyPair();
  }

  @Override
  public org.hyperledger.iroha.android.crypto.KeyGenerationOutcome generateWithOutcome(
      final String alias, final org.hyperledger.iroha.android.IrohaKeyManager.KeySecurityPreference preference)
      throws KeyManagementException {
    evictAttestations(alias);
    final KeyGenParameters effective = parametersFor(preference);
    if (effective.requireStrongBox() && !backend.metadata().strongBoxBacked()) {
      throw new KeyManagementException("StrongBox required but backend is not StrongBox-capable");
    }
    final KeyGenerationResult result = backend.generate(alias, effective);
    final org.hyperledger.iroha.android.crypto.KeyGenerationOutcome outcome =
        outcomeFor(alias, result.keyPair());
    enforcePreference(preference, outcome);
    return outcome;
  }

  /** Returns the measured security route for an existing or newly generated key. */
  @Override
  public org.hyperledger.iroha.android.crypto.KeyGenerationOutcome outcomeFor(
      final String alias, final KeyPair keyPair) throws KeyManagementException {
    final KeyProviderMetadata keyMetadata = backend.keyMetadata(alias, keyPair);
    final org.hyperledger.iroha.android.crypto.KeyGenerationOutcome.Route route;
    if (keyMetadata.strongBoxBacked()) {
      route = org.hyperledger.iroha.android.crypto.KeyGenerationOutcome.Route.STRONGBOX;
    } else if (keyMetadata.hardwareBacked()) {
      route = org.hyperledger.iroha.android.crypto.KeyGenerationOutcome.Route.HARDWARE;
    } else {
      route = org.hyperledger.iroha.android.crypto.KeyGenerationOutcome.Route.SOFTWARE;
    }
    return new org.hyperledger.iroha.android.crypto.KeyGenerationOutcome(keyPair, route);
  }

  @Override
  public KeyPair generateEphemeral() throws KeyManagementException {
    return backend.generateEphemeral(parameters);
  }

  @Override
  public boolean isHardwareBacked() {
    return backend.metadata().hardwareBacked();
  }

  @Override
  public KeyProviderMetadata metadata() {
    return backend.metadata();
  }

  @Override
  public String name() {
    return backend.name();
  }

  private static void enforcePreference(
      final org.hyperledger.iroha.android.IrohaKeyManager.KeySecurityPreference preference,
      final org.hyperledger.iroha.android.crypto.KeyGenerationOutcome outcome)
      throws KeyManagementException {
    if (preference
            == org.hyperledger.iroha.android.IrohaKeyManager.KeySecurityPreference.STRONGBOX_REQUIRED
        && outcome.route()
            != org.hyperledger.iroha.android.crypto.KeyGenerationOutcome.Route.STRONGBOX) {
      throw new KeyManagementException(
          "StrongBox required but backend produced a weaker security level");
    }
    if (preference
            == org.hyperledger.iroha.android.IrohaKeyManager.KeySecurityPreference.HARDWARE_REQUIRED
        && outcome.route()
            == org.hyperledger.iroha.android.crypto.KeyGenerationOutcome.Route.SOFTWARE) {
      throw new KeyManagementException(
          "Hardware-backed key required but backend produced a software key");
    }
  }

  /** Returns attestation material recorded for {@code alias}, when available. */
  public Optional<KeyAttestation> attestation(final String alias) {
    try {
      return fetchAttestation(alias, NO_CHALLENGE);
    } catch (final KeyManagementException ex) {
      return Optional.empty();
    }
  }

  /**
   * Requests backend-generated attestation for {@code alias}. Android Keystore cannot re-attest an
   * existing alias with a new challenge; callers must provision a new alias with the challenge in
   * {@link KeyGenParameters} when fresh evidence is required.
   */
  public Optional<KeyAttestation> generateAttestation(
      final String alias, final byte[] challenge) throws KeyManagementException {
    final byte[] normalizedChallenge = challenge == null ? NO_CHALLENGE : challenge.clone();
    final byte[] fingerprint = fingerprintChallenge(normalizedChallenge);
    if (normalizedChallenge.length == 0) {
      final Optional<KeyAttestation> cached = lookupCachedAttestation(alias, normalizedChallenge);
      if (cached.isPresent()) {
        return cached;
      }
      final Optional<KeyAttestation> existing = backend.attestation(alias);
      if (existing.isPresent()) {
        return Optional.of(cacheAttestation(alias, normalizedChallenge, existing.get()));
      }
      return backend.generateAttestation(alias, normalizedChallenge)
          .map(att -> cacheAttestation(alias, normalizedChallenge, att));
    }
    evictAttestationEntry(alias, fingerprint);
    final Optional<KeyAttestation> attestation =
        backend.generateAttestation(alias, normalizedChallenge);
    if (!attestation.isPresent()) {
      throw new KeyManagementException(
          "Attestation challenge is not supported by backend " + backend.name());
    }
    return attestation;
  }

  /**
   * Verifies attestation material (when present) using the supplied verifier.
   *
   * <p>Returns an empty optional when the backend has no attestation for {@code alias}.
   */
  @Override
  public Optional<AttestationResult> verifyAttestation(
      final String alias, final AttestationVerifier verifier, final byte[] expectedChallenge)
      throws AttestationVerificationException {
    Objects.requireNonNull(alias, "alias");
    Objects.requireNonNull(verifier, "verifier");
    if (expectedChallenge == null || expectedChallenge.length == 0) {
      throw new AttestationVerificationException(
          "Expected attestation challenge must be non-empty");
    }
    final byte[] challenge = expectedChallenge.clone();
    final Optional<KeyAttestation> attestation;
    try {
      attestation = fetchRecordedAttestation(alias);
    } catch (final KeyManagementException | RuntimeException ex) {
      throw new AttestationVerificationException(
          "Failed to obtain challenge-bound attestation material", ex);
    }
    if (!attestation.isPresent()) {
      return Optional.empty();
    }
    try {
      final AttestationResult result = verifier.verify(attestation.get(), challenge);
      final Optional<KeyPair> loadedKey;
      try {
        loadedKey = backend.load(alias);
      } catch (final KeyManagementException | RuntimeException ex) {
        throw new AttestationVerificationException(
            "Failed to load the attested alias key", ex);
      }
      if (loadedKey == null || !loadedKey.isPresent() || loadedKey.get().getPublic() == null) {
        throw new AttestationVerificationException("Attested alias key is unavailable");
      }
      final byte[] aliasPublicKey = loadedKey.get().getPublic().getEncoded();
      final java.security.PublicKey attestedLeafKey = result.leafCertificate().getPublicKey();
      final byte[] attestedPublicKey =
          attestedLeafKey == null ? null : attestedLeafKey.getEncoded();
      if (aliasPublicKey == null
          || attestedPublicKey == null
          || !MessageDigest.isEqual(aliasPublicKey, attestedPublicKey)) {
        throw new AttestationVerificationException(
            "Attested leaf public key does not match the alias key");
      }
      return Optional.of(result);
    } catch (final AttestationVerificationException ex) {
      evictAttestations(alias);
      throw ex;
    }
  }

  /** Retained for binary compatibility; always fails closed because a challenge is required. */
  @Override
  public Optional<AttestationResult> verifyAttestation(
      final String alias, final AttestationVerifier verifier)
      throws AttestationVerificationException {
    return verifyAttestation(alias, verifier, null);
  }

  /** Attempts to create a keystore-backed provider when running on Android hardware. */
  public static Optional<KeystoreKeyProvider> maybeCreate(final KeyGenParameters parameters) {
    return AndroidKeystoreBackend.maybeCreate()
        .map(backend -> new KeystoreKeyProvider(backend, parameters));
  }

  /** Returns a copy of this provider with adjusted generation parameters. */
  public KeystoreKeyProvider withParameters(final KeyGenParameters parameters) {
    return new KeystoreKeyProvider(this.backend, parameters, defaultPreference);
  }

  /** Returns a copy of this provider tuned for the requested security preference. */
  public KeystoreKeyProvider withPreference(
      final org.hyperledger.iroha.android.IrohaKeyManager.KeySecurityPreference preference) {
    return new KeystoreKeyProvider(this.backend, parametersFor(preference), preference);
  }

  private KeyGenParameters parametersFor(
      final org.hyperledger.iroha.android.IrohaKeyManager.KeySecurityPreference preference) {
    if (preference == null) {
      return this.parameters;
    }
    final KeyGenParameters.Builder builder = this.parameters.toBuilder();
    switch (preference) {
      case STRONGBOX_REQUIRED:
        builder.setRequireStrongBox(true).setPreferStrongBox(true);
        break;
      case STRONGBOX_PREFERRED:
        builder.setRequireStrongBox(false).setPreferStrongBox(true);
        break;
      case HARDWARE_REQUIRED:
        builder.setRequireStrongBox(false).setPreferStrongBox(false);
        break;
      case HARDWARE_PREFERRED:
      case SOFTWARE_ONLY:
      default:
        // leave defaults
        break;
    }
    return builder.build();
  }

  private Optional<KeyAttestation> lookupCachedAttestation(
      final String alias, final byte[] challenge) {
    final CacheKey cacheKey = new CacheKey(alias, fingerprintChallenge(challenge));
    return Optional.ofNullable(attestationCache.get(cacheKey));
  }

  private KeyAttestation cacheAttestation(
      final String alias, final byte[] challenge, final KeyAttestation attestation) {
    final CacheKey specific = new CacheKey(alias, fingerprintChallenge(challenge));
    attestationCache.put(specific, attestation);
    return attestation;
  }

  private void evictAttestations(final String alias) {
    attestationCache.keySet().removeIf(entry -> entry.alias().equals(alias));
  }

  private void evictAttestationEntry(final String alias, final byte[] challengeFingerprint) {
    final CacheKey cacheKey = new CacheKey(alias, challengeFingerprint);
    attestationCache.remove(cacheKey);
  }

  private Optional<KeyAttestation> fetchAttestation(
      final String alias, final byte[] challenge) throws KeyManagementException {
    final byte[] normalized = challenge == null ? NO_CHALLENGE : challenge.clone();
    final Optional<KeyAttestation> cached = lookupCachedAttestation(alias, normalized);
    if (cached.isPresent()) {
      return cached;
    }
    return backend.attestation(alias).map(att -> cacheAttestation(alias, normalized, att));
  }

  private Optional<KeyAttestation> fetchRecordedAttestation(final String alias)
      throws KeyManagementException {
    // Challenge-bound Android certificates are minted at key provisioning time. Always reread the
    // recorded chain here so verification cannot consume stale in-memory evidence.
    return backend.attestation(alias);
  }

  private static byte[] fingerprintChallenge(final byte[] challenge) {
    if (challenge == null || challenge.length == 0) {
      return NO_CHALLENGE;
    }
    return Arrays.copyOf(challenge, challenge.length);
  }

  private record CacheKey(String alias, byte[] challengeFingerprint) {
    private CacheKey(final String alias, final byte[] challengeFingerprint) {
      if (alias == null) {
        throw new NullPointerException("alias");
      }
      if (alias.trim().isEmpty()) {
        throw new IllegalArgumentException("alias must not be blank");
      }
      if (challengeFingerprint == null) {
        throw new NullPointerException("challengeFingerprint");
      }
      this.alias = alias;
      this.challengeFingerprint = Arrays.copyOf(challengeFingerprint, challengeFingerprint.length);
    }

    @Override
    public int hashCode() {
      int result = alias.hashCode();
      result = 31 * result + Arrays.hashCode(challengeFingerprint);
      return result;
    }

    @Override
    public boolean equals(final Object obj) {
      if (this == obj) {
        return true;
      }
      if (!(obj instanceof CacheKey other)) {
        return false;
      }
      return alias.equals(other.alias)
          && Arrays.equals(challengeFingerprint, other.challengeFingerprint);
    }
  }
}

File: kotlin/client-android/src/main/java/org/hyperledger/iroha/sdk/crypto/keystore/KeystoreKeyProvider.kt
package org.hyperledger.iroha.sdk.crypto.keystore

import java.security.KeyPair
import java.util.concurrent.ConcurrentHashMap
import org.hyperledger.iroha.sdk.crypto.KeyGenerationOutcome
import org.hyperledger.iroha.sdk.crypto.KeyManagementException
import org.hyperledger.iroha.sdk.crypto.KeyProvider
import org.hyperledger.iroha.sdk.crypto.KeyProviderMetadata
import org.hyperledger.iroha.sdk.crypto.keystore.attestation.AttestationResult
import org.hyperledger.iroha.sdk.crypto.keystore.attestation.AttestationVerificationException
import org.hyperledger.iroha.sdk.crypto.keystore.attestation.AttestationVerifier

/**
 * `KeyProvider` backed by an Android Keystore-compatible backend.
 *
 * The backend is supplied through `KeystoreBackend`; desktop JVM builds can rely on fake
 * implementations for tests while Android builds will provide a real backend that bridges to
 * `android.security.keystore.KeyStore` / `KeyGenParameterSpec`. Support for StrongBox
 * and discrete secure elements will land by swapping the backend at runtime depending on device
 * capabilities.
 */
class KeystoreKeyProvider(
    private val backend: KeystoreBackend,
    private val parameters: KeyGenParameters,
) : KeyProvider {

    private val attestationCache = ConcurrentHashMap<CacheKey, KeyAttestation>()

    @Throws(KeyManagementException::class)
    override fun load(alias: String): KeyPair? = backend.load(alias)

    @Throws(KeyManagementException::class)
    override fun generate(alias: String): KeyPair {
        evictAttestations(alias)
        val result = backend.generate(alias, parameters)
        if (parameters.requireStrongBox
            && !result.strongBoxBacked
            && !backend.metadata().strongBoxBacked
        ) {
            throw KeyManagementException("StrongBox required but backend is not StrongBox-capable")
        }
        return result.keyPair
    }

    @Throws(KeyManagementException::class)
    fun generate(alias: String, preference: KeySecurityPreference): KeyPair {
        evictAttestations(alias)
        val effective = parametersFor(preference)
        if (effective.requireStrongBox && !backend.metadata().strongBoxBacked) {
            throw KeyManagementException("StrongBox required but backend is not StrongBox-capable")
        }
        val result = backend.generate(alias, effective)
        if (preference == KeySecurityPreference.STRONGBOX_REQUIRED && !result.strongBoxBacked) {
            throw KeyManagementException(
                "StrongBox required but backend fell back to a weaker security level"
            )
        }
        return result.keyPair
    }

    @Throws(KeyManagementException::class)
    fun generateWithOutcome(
        alias: String,
        preference: KeySecurityPreference,
    ): KeyGenerationOutcome {
        evictAttestations(alias)
        val effective = parametersFor(preference)
        val result = backend.generate(alias, effective)
        val route = when {
            result.strongBoxBacked -> KeyGenerationOutcome.Route.STRONGBOX
            backend.metadata().hardwareBacked -> KeyGenerationOutcome.Route.HARDWARE
            else -> KeyGenerationOutcome.Route.SOFTWARE
        }
        if (preference == KeySecurityPreference.STRONGBOX_REQUIRED
            && route != KeyGenerationOutcome.Route.STRONGBOX
        ) {
            throw KeyManagementException(
                "StrongBox required but backend fell back to a weaker security level"
            )
        }
        return KeyGenerationOutcome(result.keyPair, route)
    }

    @Throws(KeyManagementException::class)
    override fun generateEphemeral(): KeyPair = backend.generateEphemeral(parameters)

    override fun isHardwareBacked(): Boolean = backend.metadata().hardwareBacked

    override fun metadata(): KeyProviderMetadata = backend.metadata()

    override fun name(): String = backend.name()

    /** Returns attestation material recorded for `alias`, when available. */
    fun attestation(alias: String): KeyAttestation? =
        try {
            fetchAttestation(alias, NO_CHALLENGE)
        } catch (_: KeyManagementException) {
            null
        }

    /**
     * Requests fresh attestation for `alias`. Providers that do not support attestation return
     * null.
     */
    @Throws(KeyManagementException::class)
    fun generateAttestation(alias: String, challenge: ByteArray?): KeyAttestation? {
        val normalizedChallenge = challenge?.copyOf() ?: NO_CHALLENGE
        val fingerprint = fingerprintChallenge(normalizedChallenge)
        if (normalizedChallenge.isEmpty()) {
            val cached = lookupCachedAttestation(alias, normalizedChallenge)
            if (cached != null) return cached
            val existing = backend.attestation(alias)
            if (existing != null) return cacheAttestation(alias, normalizedChallenge, existing)
            return backend.generateAttestation(alias, normalizedChallenge)
                ?.let { cacheAttestation(alias, normalizedChallenge, it) }
        }
        evictAttestationEntry(alias, fingerprint)
        val attestation = backend.generateAttestation(alias, normalizedChallenge)
            ?: throw KeyManagementException(
                "Attestation challenge is not supported by backend ${backend.name()}"
            )
        return cacheAttestation(alias, normalizedChallenge, attestation)
    }

    /**
     * Verifies attestation material (when present) using the supplied verifier.
     *
     * Returns null when the backend has no attestation for `alias`.
     */
    @Throws(AttestationVerificationException::class)
    fun verifyAttestation(
        alias: String,
        verifier: AttestationVerifier,
        expectedChallenge: ByteArray?,
    ): AttestationResult? {
        val challenge = expectedChallenge?.copyOf() ?: NO_CHALLENGE
        val attestation = try {
            fetchAttestation(alias, challenge)
        } catch (_: KeyManagementException) {
            return null
        } ?: return null
        try {
            return if (expectedChallenge == null) {
                verifier.verify(attestation)
            } else {
                verifier.verify(attestation, expectedChallenge.copyOf())
            }
        } catch (ex: AttestationVerificationException) {
            evictAttestations(alias)
            throw ex
        }
    }

    /** Verifies attestation without enforcing a challenge match. */
    @Throws(AttestationVerificationException::class)
    fun verifyAttestation(
        alias: String,
        verifier: AttestationVerifier,
    ): AttestationResult? = verifyAttestation(alias, verifier, null)

    /** Returns a copy of this provider with adjusted generation parameters. */
    fun withParameters(parameters: KeyGenParameters): KeystoreKeyProvider =
        KeystoreKeyProvider(backend, parameters)

    /** Returns a copy of this provider tuned for the requested security preference. */
    fun withPreference(preference: KeySecurityPreference): KeystoreKeyProvider =
        KeystoreKeyProvider(backend, parametersFor(preference))

    private fun parametersFor(preference: KeySecurityPreference?): KeyGenParameters {
        if (preference == null) return parameters
        val builder = parameters.toBuilder()
        when (preference) {
            KeySecurityPreference.STRONGBOX_REQUIRED ->
                builder.setRequireStrongBox(true)
                    .setPreferStrongBox(true)
                    .setAllowStrongBoxFallback(false)
            KeySecurityPreference.STRONGBOX_PREFERRED ->
                builder.setRequireStrongBox(false)
                    .setPreferStrongBox(true)
                    .setAllowStrongBoxFallback(true)
            KeySecurityPreference.HARDWARE_REQUIRED ->
                builder.setRequireStrongBox(false)
                    .setPreferStrongBox(false)
                    .setAllowStrongBoxFallback(true)
            KeySecurityPreference.HARDWARE_PREFERRED,
            KeySecurityPreference.SOFTWARE_ONLY -> { /* leave defaults */ }
        }
        return builder.build()
    }

    private fun lookupCachedAttestation(alias: String, challenge: ByteArray): KeyAttestation? {
        val cacheKey = CacheKey(alias, fingerprintChallenge(challenge))
        return attestationCache[cacheKey]
    }

    private fun cacheAttestation(
        alias: String,
        challenge: ByteArray,
        attestation: KeyAttestation,
    ): KeyAttestation {
        val specific = CacheKey(alias, fingerprintChallenge(challenge))
        attestationCache[specific] = attestation
        return attestation
    }

    private fun evictAttestations(alias: String) {
        attestationCache.keys.removeAll { it.alias == alias }
    }

    private fun evictAttestationEntry(alias: String, challengeFingerprint: ByteArray) {
        attestationCache.remove(CacheKey(alias, challengeFingerprint))
    }

    private fun fetchAttestation(alias: String, challenge: ByteArray?): KeyAttestation? {
        val normalized = challenge?.copyOf() ?: NO_CHALLENGE
        val cached = lookupCachedAttestation(alias, normalized)
        if (cached != null) return cached
        if (normalized.isNotEmpty()) {
            val generated = backend.generateAttestation(alias, normalized)
            if (generated != null) return cacheAttestation(alias, normalized, generated)
        }
        return backend.attestation(alias)?.let { cacheAttestation(alias, normalized, it) }
    }

    private class CacheKey(val alias: String, challengeFingerprint: ByteArray) {
        private val _challengeFingerprint: ByteArray

        init {
            require(alias.isNotBlank()) { "alias must not be blank" }
            _challengeFingerprint = challengeFingerprint.copyOf()
        }

        override fun hashCode(): Int {
            var result = alias.hashCode()
            result = 31 * result + _challengeFingerprint.contentHashCode()
            return result
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is CacheKey) return false
            return alias == other.alias
                && _challengeFingerprint.contentEquals(other._challengeFingerprint)
        }
    }

    companion object {
        private val NO_CHALLENGE = ByteArray(0)

        /** Attempts to create a keystore-backed provider when running on Android hardware. */
        @JvmStatic
        fun maybeCreate(parameters: KeyGenParameters): KeystoreKeyProvider? =
            AndroidKeystoreBackend.maybeCreate()?.let { KeystoreKeyProvider(it, parameters) }

        private fun fingerprintChallenge(challenge: ByteArray?): ByteArray {
            if (challenge == null || challenge.isEmpty()) return NO_CHALLENGE
            return challenge.copyOf()
        }
    }
}
